File: src/main/java/com/github/svegon/capi/argument_types/MappedArgumentType.java
package com.github.svegon.capi.argument_types;

import com.mojang.brigadier.StringReader;
import com.mojang.brigadier.arguments.ArgumentType;
import com.mojang.brigadier.context.CommandContext;
import com.mojang.brigadier.exceptions.CommandSyntaxException;
import com.mojang.brigadier.suggestion.Suggestions;
import com.mojang.brigadier.suggestion.SuggestionsBuilder;
import net.minecraft.command.CommandSource;

import java.util.Collection;
import java.util.concurrent.CompletableFuture;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.Supplier;

public class MappedArgumentType<T, U> implements ArgumentType<U> {
    private final ArgumentType<T> model;
    private final Function<? super T, ? extends U> mapper;
    private final BiFunction<? super CommandContext<? extends CommandSource>, ? super SuggestionsBuilder,
            ? extends CompletableFuture<Suggestions>> suggestionCompleter;
    private final Collection<String> examples;

    private MappedArgumentType(ArgumentType<T> model, Function<? super T, ? extends U> mapper,
                               BiFunction<? super CommandContext<? extends CommandSource>, ? super SuggestionsBuilder,
                                       ? extends CompletableFuture<Suggestions>> suggestionCompleter,
                               Collection<String> examples) {
        this.model = model;
        this.mapper = mapper;
        this.suggestionCompleter = suggestionCompleter;
        this.examples = examples;
    }

    @Override
    public U parse(StringReader reader) throws CommandSyntaxException {
        return mapper.apply(model.parse(reader));
    }

    @Override
    @SuppressWarnings("unchecked")
    public <S> CompletableFuture<Suggestions> listSuggestions(CommandContext<S> context, SuggestionsBuilder builder) {
        return suggestionCompleter.apply((CommandContext<CommandSource>) context, builder);
    }

    @Override
    public Collection<String> getExamples() {
        return examples;
    }

    public static <T, U> MappedArgumentType<T, U> of(ArgumentType<T> original,
                                                     Function<? super T, ? extends U> mapper,
                                                     BiFunction<? super CommandContext<? extends CommandSource>,
                                                             ? super SuggestionsBuilder,
                                                         ? extends CompletableFuture<Suggestions>> suggestionCompleter,
                                                     Collection<String> examples) {
        return new MappedArgumentType<>(original, mapper, suggestionCompleter, examples);
    }

    public static <T, U> MappedArgumentType<T, U> of(ArgumentType<T> original,
                                                     Function<? super T, ? extends U> mapper,
                                                     Collection<String> examples) {
        return of(original, mapper, original::listSuggestions, examples);
    }

    public static <T, U> MappedArgumentType<T, U> of(ArgumentType<T> original,
                                                     Function<? super T, ? extends U> mapper) {
        return of(original, mapper, original::listSuggestions, original.getExamples());
    }

    public static <T, U> MappedArgumentType<T, U> of(Supplier<ArgumentType<T>> originalSupplier,
                                                     Function<? super T, ? extends U> mapper) {
        return of(originalSupplier.get(), mapper);
    }
}

File: src/main/kotlin/io/github/svegon/mclientapi/argument_types/MappedArgumentType.kt
package io.github.svegon.capi.argument_types

import com.mojang.brigadier.StringReader
import com.mojang.brigadier.arguments.ArgumentType
import com.mojang.brigadier.context.CommandContext
import com.mojang.brigadier.exceptions.CommandSyntaxException
import com.mojang.brigadier.suggestion.Suggestions
import com.mojang.brigadier.suggestion.SuggestionsBuilder
import net.minecraft.command.CommandSource
import java.util.concurrent.CompletableFuture
import java.util.function.BiFunction
import java.util.function.Function
import java.util.function.Supplier

class MappedArgumentType<T, U> private constructor(
    private val model: ArgumentType<T>, private val mapper: Function<in T, out U>,
    private val suggestionCompleter: BiFunction<in CommandContext<out CommandSource>, in SuggestionsBuilder, out CompletableFuture<Suggestions>>,
    private val examples: Collection<String>
) : ArgumentType<U> {
    @Throws(CommandSyntaxException::class)
    override fun parse(reader: StringReader): U {
        return mapper.apply(model.parse(reader))
    }

    override fun <S> listSuggestions(
        context: CommandContext<S>,
        builder: SuggestionsBuilder
    ): CompletableFuture<Suggestions> {
        return suggestionCompleter.apply(context as CommandContext<CommandSource>, builder)
    }

    override fun getExamples(): Collection<String> {
        return examples
    }

    companion object {
        fun <T, U> of(
            original: ArgumentType<T>,
            mapper: Function<in T, out U>,
            suggestionCompleter: BiFunction<in CommandContext<out CommandSource>, in SuggestionsBuilder, out CompletableFuture<Suggestions>>,
            examples: Collection<String>
        ): MappedArgumentType<T, U> {
            return MappedArgumentType(original, mapper, suggestionCompleter, examples)
        }

        fun <T, U> of(
            original: ArgumentType<T>,
            mapper: Function<in T, out U>,
            examples: Collection<String>
        ): MappedArgumentType<T, U> {
            return of(
                original,
                mapper,
                { context: CommandContext<out CommandSource>?, builder: SuggestionsBuilder? ->
                    original.listSuggestions(
                        context,
                        builder
                    )
                },
                examples
            )
        }

        fun <T, U> of(
            original: ArgumentType<T>,
            mapper: Function<in T, out U>
        ): MappedArgumentType<T, U> {
            return of(
                original,
                mapper,
                { context: CommandContext<out CommandSource>?, builder: SuggestionsBuilder? ->
                    original.listSuggestions(
                        context,
                        builder
                    )
                },
                original.examples
            )
        }

        fun <T, U> of(
            originalSupplier: Supplier<ArgumentType<T>>,
            mapper: Function<in T, out U>
        ): MappedArgumentType<T, U> {
            return of(originalSupplier.get(), mapper)
        }
    }
}
